File: src/scala/src/test/scala/dealing/with/generalization/extractInterface.scala
package dealing.`with`.generalization

import org.scalamock.scalatest.MockFactory
import org.scalatest.funsuite.AnyFunSuite

class extractInterface extends AnyFunSuite with MockFactory {
  private val console = mock[BusinessLogger]
  private val triAthlete = TriAthlete("Alistair Brownlee", console)
  private val jumpyAthlete = JumpyAthlete("A crazy guy", console)

  test("triAthlete can swim") {
    expectCall("Alistair Brownlee started swimming")
    triAthlete.swim()
  }

  test("triAthlete can cycle") {
    expectCall("Alistair Brownlee started cycling")
    triAthlete.cycle()
  }

  test("triAthlete can run") {
    expectCall("Alistair Brownlee started running")
    triAthlete.run()
  }

  test("jumpyAthlete can swim") {
    expectCall("A crazy guy started swimming")
    jumpyAthlete.swim()
  }

  test("jumpyAthlete can cycle") {
    expectCall("A crazy guy started cycling")
    jumpyAthlete.cycle()
  }

  test("jumpyAthlete can run") {
    expectCall("A crazy guy started running")
    jumpyAthlete.run()
  }

  test("jumpyAthlete can skydive") {
    expectCall("A crazy guy is skydiving")
    jumpyAthlete.jump()
  }

  private def expectCall(expected: String): Unit = {
    (console.log _).expects(expected).once()

  }
}

File: src/java/test/java/dealing/with/generalization/extractInterface.java
package dealing.with.generalization;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;

import java.io.ByteArrayOutputStream;
import java.io.PrintStream;

import static org.assertj.core.api.Assertions.assertThat;

class extractInterface {
    private final ByteArrayOutputStream console = new ByteArrayOutputStream();

    @BeforeEach
    void init() {
        System.setOut(new PrintStream(console));
    }

    private void assertConsoleOut(String expected) {
        assertThat(console.toString()).isEqualToIgnoringNewLines(expected);
    }

    @Nested
    class TriAthlete_should {
        private final TriAthlete triAthlete = new TriAthlete("Alistair Brownlee");

        @Test
        void swim() {
            triAthlete.swim();
            assertConsoleOut("Alistair Brownlee started swimming");
        }

        @Test
        void cycle() {
            triAthlete.cycle();
            assertConsoleOut("Alistair Brownlee started cycling");
        }

        @Test
        void run() {
            triAthlete.run();
            assertConsoleOut("Alistair Brownlee started running");
        }
    }

    @Nested
    class JumpyAthlete_should {
        private final JumpyAthlete jumpyAthlete = new JumpyAthlete("A crazy guy");

        @Test
        void swim() {
            jumpyAthlete.swim();
            assertConsoleOut("A crazy guy started swimming");
        }

        @Test
        void cycle() {
            jumpyAthlete.cycle();
            assertConsoleOut("A crazy guy started cycling");
        }

        @Test
        void jump() {
            jumpyAthlete.jump();
            assertConsoleOut("A crazy guy is skydiving");
        }

        @Test
        void run() {
            jumpyAthlete.run();
            assertConsoleOut("A crazy guy started running");
        }
    }
}
